src/frontend/js/components/PaymentButton/index.tsx

```diff
@@ -0,0 +1,226 @@
+    defaultMessage: 'Payment in progress',
+    description: 'Label for screen reader when a payment is in progress.',
+    id: 'components.PaymentButton.paymentInProgress',
+  },
+});
+
+export enum PaymentErrorMessageId {
+  ERROR_DEFAULT = 'errorDefault',
+  ERROR_ABORT = 'errorAbort',
+  ERROR_ABORTING = 'errorAborting',
+}
+
+interface PaymentButtonProps {
+  product: Joanie.Product;
+  billingAddress?: Joanie.Address;
+  creditCard?: Nullable<Joanie.CreditCard['id']>;
+  onSuccess: () => void;
+}
+
+type PaymentInfo = Joanie.Payment & { order_id: string };
+type OneClickPaymentInfo = Joanie.PaymentOneClick & { order_id: string };
+
+enum ComponentStates {
+  IDLE = 'idle',
+  LOADING = 'loading',
+  ERROR = 'error',
+}
+
+/**
+ * Displays a button to proceed to the payment.
+ * First it creates the payment from Joanie then displays the payment interface
+ * or the error message.
+ */
+const PaymentButton = ({ product, billingAddress, creditCard, onSuccess }: PaymentButtonProps) => {
+  const intl = useIntl();
+  const API = useJoanieApi();
+  const timeoutRef = useRef<NodeJS.Timeout>();
+  const courseCode = useCourseCode();
+  const orderManager = useOrders();
+
+  const isReadyToPay = useMemo(() => {
+    return courseCode && product.id && billingAddress;
+  }, [product, courseCode, billingAddress]);
+
+  const [payment, setPayment] = useState<PaymentInfo | OneClickPaymentInfo>();
+  const [state, setState] = useState<ComponentStates>(ComponentStates.IDLE);
+  const [error, setError] = useState<PaymentErrorMessageId>(PaymentErrorMessageId.ERROR_DEFAULT);
+
+  /**
+   * Use Joanie API to retrieve an order and check if it's state is validated
+   *
+   * @param {string} id - Order id
+   * @returns {Promise<boolean>} - Promise resolving to true if order is validated
+   */
+  const isOrderValidated = async (id: string): Promise<Boolean> => {
+    const order = await API.user.orders.get(id);
+    return order?.state === OrderState.VALIDATED;
+  };
+
+  /** type guard to check if the payment is a payment one click */
+  const isOneClickPayment = (p: typeof payment): p is OneClickPaymentInfo =>
+    (p as OneClickPaymentInfo)?.is_paid === true;
+
+  const createPayment = async () => {
+    if (isReadyToPay) {
+      setState(ComponentStates.LOADING);
+      let paymentInfos = payment;
+
+      if (!paymentInfos) {
+        try {
+          const order = await orderManager.methods.create({
+            billing_address: billingAddress!,
+            ...(creditCard && { credit_card_id: creditCard }),
+            course: courseCode,
+            product: product.id,
+          });
+          paymentInfos = {
+            ...order.payment_info!,
+            order_id: order.id,
+          };
+        } catch {
+          setState(ComponentStates.ERROR);
+        }
+      }
+
+      setPayment(paymentInfos);
+    }
+  };
+
+  const handleSuccess = () => {
+    let round = 0;
+
+    const checkOrderValidity = async () => {
+      if (round >= 30) {
```

Comment:
```suggestion
      if (round >= 30) {
```
can we add const for this so it's easily configurable (to be used in the test as well)

Reply:
Oh yes ! Good catch, this is something I forgot to do.